dns/query.py

```diff
@@ -290,19 +301,47 @@ def https(q, where, timeout=None, port=443, source=None, source_port=0,
         elif af == socket.AF_INET6:
             url = 'https://[{}]:{}{}'.format(where, port, path)
     elif bootstrap_address is not None:
+        _httpx_ok = False
         split_url = urllib.parse.urlsplit(where)
         headers['Host'] = split_url.hostname
         url = where.replace(split_url.hostname, bootstrap_address)
-        transport_adapter = HostHeaderSSLAdapter()
+        if _have_requests:
+            transport_adapter = HostHeaderSSLAdapter()
     else:
         url = where
     if source is not None:
         # set source port and source address
-        transport_adapter = SourceAddressAdapter(source)
+        if _have_httpx:
+            if source_port == 0:
+                transport = httpx.HTTPTransport(local_address=source[0])
+            else:
+                _httpx_ok = False
+        if _have_requests:
+            transport_adapter = SourceAddressAdapter(source)
+
+    if not _httpx_ok and not _have_requests:
+        raise NoDOH('Cannot use httpx for this operation, and '
+                    'requests is not available.')
 
     with contextlib.ExitStack() as stack:
+        if session:
+            if _have_httpx:
+                _is_httpx = isinstance(session, httpx.Client)
+            else:
+                _is_httpx = False
+            if _is_httpx and not _httpx_ok:
+                # we can't use this session
+                session = None
```

Comment:
I'd probably move this check above the code that raises; if the user passes in an httpx.Client, but httpx can't be used, it would be better to raise an exception than silently ignore the session.